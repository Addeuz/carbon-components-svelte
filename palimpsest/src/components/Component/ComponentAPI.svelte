<script>
  export let props = [];

  import {
    StructuredList,
    StructuredListBody,
    StructuredListHead,
    StructuredListRow,
    StructuredListCell,
    Tag,
    CodeSnippet
  } from 'carbon-components-svelte';
  import copy from 'clipboard-copy';

  const propTypeColor = {
    string: 'green',
    number: 'blue',
    boolean: 'purple'
  };

  function formatValue(value) {
    return typeof value === 'string' ? `"${value}"` : value;
  }
</script>

<style>
  .bx--row {
    margin-top: 1rem;
  }

  .values {
    display: flex;
  }

  :global(.value) {
    margin-right: 0.5rem;
  }

  :global(.value:not(:first-of-type)) {
    margin-left: 0.5rem;
  }

  .separator {
    color: var(--cds-text-03);
  }
</style>

<div class="bx--row">
  <StructuredList>
    <StructuredListHead>
      <StructuredListRow head>
        <StructuredListCell head>Prop</StructuredListCell>
        <StructuredListCell head>Type</StructuredListCell>
        <StructuredListCell head>Value</StructuredListCell>
        <StructuredListCell head>Description</StructuredListCell>
      </StructuredListRow>
    </StructuredListHead>
    <StructuredListBody>
      {#each props as prop, i (prop.name)}
        <StructuredListRow>
          <StructuredListCell noWrap>{prop.name}</StructuredListCell>
          <StructuredListCell noWrap>
            <Tag type={propTypeColor[prop.type]}>
              <code>{prop.type}</code>
            </Tag>
          </StructuredListCell>
          <StructuredListCell>
            <div class="values">
              {#each prop.value as value, j (value)}
                <CodeSnippet
                  class="value"
                  type="inline"
                  code={formatValue(value)}
                  on:click={() => {
                    copy(formatValue(value));
                  }} />
                {#if j < prop.value.length - 1}
                  <span class="separator">|</span>
                {/if}
              {/each}
            </div>
            {#if prop.defaultValue}
              <div style="margin-top: .75rem">
                <strong>Default:</strong>
                <code>{formatValue(prop.defaultValue.value)}</code>
              </div>
            {/if}
          </StructuredListCell>
          <StructuredListCell>
            {@html prop.description}
          </StructuredListCell>
        </StructuredListRow>
      {/each}
    </StructuredListBody>
  </StructuredList>
</div>
