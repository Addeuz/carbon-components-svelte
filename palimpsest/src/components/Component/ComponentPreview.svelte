<script>
  export let minHeight = '11.5rem';
  export let light = false;
  export let code = '';

  import copy from 'clipboard-copy';
  import { CodeSnippet } from 'carbon-components-svelte';
  import { theme } from '../../store';

  $: code = code.trim().replace(/< \//g, '</');
  $: light = light && ['white', 'g10'].includes($theme);
</script>

<style>
  .preview {
    position: relative;
    z-index: 1;
    border: 1px solid var(--cds-ui-03);
  }

  .preview.light {
    background-color: var(--cds-ui-01);
  }

  .grid {
    position: absolute;
    z-index: -1;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
  }

  .grid--row {
    display: flex;
    height: 100%;
  }

  .grid--col:not(:last-of-type) {
    border-right: 1px dashed var(--cds-ui-03);
  }

  .grid--col.light {
    background-color: var(--cds-ui-01);
  }

  .grid--col--outer {
    display: flex;
    height: 100%;
  }

  .grid--col--inner {
    flex: 1;
    background-color: var(--cds-ui-background);
  }

  .grid--col--inner {
    border-left: 1px dashed var(--cds-ui-03);
    border-right: 1px dashed var(--cds-ui-03);
  }

  .grid--col--inner.light {
    background-color: var(--cds-ui-01);
  }

  h3 {
    margin-top: 2rem;
    margin-bottom: 1.5rem;
    font-size: 1.5rem;
  }

  .wrapper {
    display: inline-flex;
    flex: 1;
    width: 100%;
    min-height: 9rem;
    padding: 2rem 0 2rem 1rem;
  }
</style>

<div class="preview bx--row" class:light>
  <div class="grid">
    <div class="grid--row">
      {#each Array.from({ length: 8 }, (_, i) => i) as item, i (item)}
        <div class="grid--col bx--col" class:light>
          <div class="grid--col--outer">
            <div class="grid--col--inner" class:light />
          </div>
        </div>
      {/each}
    </div>
  </div>
  <div class="bx--col bx--no-gutter--left">
    <div class="wrapper" style={`min-height: ${minHeight}`}>
      <slot />
    </div>
  </div>
</div>

<div class="bx--row">
  <div class="bx--col-lg-4">
    <h3>Props</h3>
    <slot name="props" />
  </div>
  <div class="bx--col">
    <h3>Code</h3>
    <CodeSnippet type="multi" on:click={() => copy(code)} {code} />
    <slot name="code" />
  </div>
</div>
