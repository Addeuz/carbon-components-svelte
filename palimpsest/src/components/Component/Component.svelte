<script context="module">
  import Accordion from '../carbon-components/Accordion';
  import CodeSnippet from '../carbon-components/CodeSnippet';

  export const components = {
    Accordion,
    CodeSnippet
  };
</script>

<script>
  import { afterUpdate, setContext } from 'svelte';
  import { Tabs, Tab } from 'carbon-components-svelte';
  import { push, location } from 'svelte-spa-router';
  import Router from 'svelte-spa-router';
  import ComponentTab from './ComponentTab.svelte';

  const prefix = '/c';
  const urls = [{ path: '' }, { path: '/API' }, { path: '/kitchen-sink' }];

  let selected = 0;
  let component = undefined;
  let prevComponent = undefined;

  setContext('Component', {
    set: name => {
      component = name;
    }
  });

  afterUpdate(() => {
    if (component !== prevComponent && prevComponent !== undefined) {
      window.scrollTo(0, 0);
    }

    prevComponent = component;
  });

  $: baseUrl = component ? `#${prefix}/${component}` : $location;
  $: {
    switch (
      $location
        .split('/')
        .pop()
        .toLowerCase()
    ) {
      case 'api':
        selected = 1;
        break;
      case 'kitchen-sink':
        selected = 2;
        break;
      default:
        selected = 0;
        break;
    }
  }
</script>

<style>
  main {
    margin-top: 5rem;
    padding-left: 12rem;
    padding-bottom: 2.5rem;
  }

  @media screen and (max-width: 768px) {
    main {
      padding-left: 0;
    }
  }

  h1 {
    margin-bottom: 1.5rem;
  }

  .bx--grid {
    max-width: 66rem;
  }

  :global(body) {
    overflow-y: scroll;
  }
</style>

<main>
  <div class="bx--grid">
    <header class="bx--row">
      <div class="bx--col">
        <h1>{component}</h1>
      </div>
    </header>
    <div class="bx--row">
      <div class="bx--col bx--no-gutter">
        <Tabs
          type="container"
          bind:selected
          on:change={({ detail }) => {
            push(`${prefix}/${component}${urls[detail].path}`);
          }}>
          <Tab label="Preview" href={baseUrl} />
          <Tab label="API" href={`${baseUrl}/API`} />
          <!-- <Tab disabled label="Kitchen Sink" href={`${baseUrl}/kitchen-sink`} /> -->
        </Tabs>
      </div>
    </div>
  </div>
  <section class="bx--grid">
    <Router {prefix} routes={{ '/:name': ComponentTab, '/:name/:tab': ComponentTab }} />
  </section>
</main>
