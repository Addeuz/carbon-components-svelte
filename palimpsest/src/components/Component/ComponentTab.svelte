<script>
  export let params = {};

  import { getContext } from 'svelte';
  import { components } from './Component.svelte';

  const { set } = getContext('Component');

  const paramMap = {
    Preview: 'Preview',
    API: 'API',
    'kitchen-sink': 'KitchenSink'
  };

  $: tabName = params.tab || 'Preview';
  $: tab = paramMap[tabName];
  $: {
    set(params.name);
    document.title = `${params.name} Â· ${tabName}`;
  }
</script>

<svelte:component this={components[params.name][tab]} />
