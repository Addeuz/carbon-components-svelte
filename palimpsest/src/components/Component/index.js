import Component from './Component.svelte';

export default Component;
export { default as ComponentPreview } from './ComponentPreview.svelte';
export { default as ComponentAPI } from './ComponentAPI.svelte';
