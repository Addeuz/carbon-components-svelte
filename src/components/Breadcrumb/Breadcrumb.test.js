import { fireEvent, render } from '@testing-library/svelte';
import Component from './Breadcrumb.Story.svelte';

describe('Breadcrumb', () => {
  function getLastBreadcrumbItem(container) {
    const breadcrumbItems = container.querySelectorAll('.bx--breadcrumb-item');
    return breadcrumbItems[breadcrumbItems.length - 1];
  }

  test('default', () => {
    const { getByText, container, rerender } = render(Component, { noTrailingSlash: false });
    const selector = '.bx--breadcrumb--no-trailing-slash';
    expect(getByText('Breadcrumb 1')).toBeInTheDocument();
    expect(getByText('Breadcrumb 2')).toBeInTheDocument();
    expect(getByText('Breadcrumb 3')).toBeInTheDocument();
    expect(container.querySelector(selector)).not.toBeInTheDocument();
    expect(container.innerHTML).toMatchSnapshot();

    rerender({ props: { noTrailingSlash: true } });
    expect(container.querySelector(selector)).toBeInTheDocument();
    expect(container.innerHTML).toMatchSnapshot();
  });

  test('skeleton', () => {
    const { container } = render(Component, { story: 'skeleton' });
    expect(container.innerHTML).toMatchSnapshot();
  });

  test('current page', () => {
    const { container } = render(Component, { story: 'current page' });
    const lastItem = getLastBreadcrumbItem(container);
    expect(lastItem.classList.contains('bx--breadcrumb-item--current')).toEqual(true);
  });

  test('current page with aria-current', () => {
    const { container } = render(Component, { story: 'current page with aria-current' });
    const lastItem = getLastBreadcrumbItem(container);
    expect(lastItem.querySelector('[aria-current="page"]')).toBeTruthy();
  });
});
