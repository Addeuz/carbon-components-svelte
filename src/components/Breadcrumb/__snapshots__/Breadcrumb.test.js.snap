// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumb default 1`] = `"<div><div class=\\"layout\\"><nav aria-label=\\"Breadcrumb\\"><ol class=\\"bx--breadcrumb\\"><li class=\\"bx--breadcrumb-item\\"><a class=\\"bx--link\\" href=\\"/#\\">Breadcrumb 1</a></li> <li class=\\"bx--breadcrumb-item\\"><a class=\\"bx--link\\" href=\\"#\\">Breadcrumb 2</a></li> <li class=\\"bx--breadcrumb-item\\"><a class=\\"bx--link\\" href=\\"#\\">Breadcrumb 3</a></li></ol></nav></div></div>"`;

exports[`Breadcrumb default 2`] = `"<div><div class=\\"layout\\"><nav aria-label=\\"Breadcrumb\\"><ol class=\\"bx--breadcrumb bx--breadcrumb--no-trailing-slash\\"><li class=\\"bx--breadcrumb-item\\"><a class=\\"bx--link\\" href=\\"/#\\">Breadcrumb 1</a></li> <li class=\\"bx--breadcrumb-item\\"><a class=\\"bx--link\\" href=\\"#\\">Breadcrumb 2</a></li> <li class=\\"bx--breadcrumb-item\\"><a class=\\"bx--link\\" href=\\"#\\">Breadcrumb 3</a></li></ol></nav></div></div>"`;

exports[`Breadcrumb skeleton 1`] = `"<div><div class=\\"layout\\"><div class=\\"bx--breadcrumb bx--skeleton\\"><div class=\\"bx--breadcrumb-item\\"><span class=\\"bx--link\\">&nbsp;</span> </div><div class=\\"bx--breadcrumb-item\\"><span class=\\"bx--link\\">&nbsp;</span> </div><div class=\\"bx--breadcrumb-item\\"><span class=\\"bx--link\\">&nbsp;</span> </div></div></div></div>"`;
